Accounts

jeisses opened the pull request.

jeisses committed:
Return transaction result in <p-should-succeed! macros

jeisses committed:
Add beginning of account contract (wip)

jeisses committed:
Add test for checking opened balances

jeisses committed:
Add despoit and transfer actions to account system (wip)

jeisses committed:
Add account nonce and fix ripemd160 address hashing

the account none is necessary to avoid replay attacks

jeisses committed:
Ensure user can't have multiple balances for the same symbol

jeisses committed:
Do not error when opening a double balance

it's convenient to let the action succeed with no changes, just like eosio.token

jeisses committed:
Add readme for account contract

jeisses committed:
Add withdraw action to account contract

jeisses committed:
Add distinction between virtual account message types

so that message / signature collisions can not happen between different actions

jeisses committed:
Refactor account actions to vaccount, add vtransfer action notify

laurensV commented on contracts/network/network.cpp:
```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+void network::transfer_handler(name from, name to, asset quantity, std::string memo) {
```
Add transfer_handler functionality (deposit into your account when memo is your account id) to README

laurensV commented on contracts/network/README.md:
````diff
@@ -0,0 +1,48 @@
+#### transfer
+Transfer tokens between virtual accounts
+```
+transfer(uint64_t from_id,
````
function is named `vtransfer`

jeisses committed:
Update account contract readme

jeisses committed:
Refactor network smart contract to vaccount

the account system will reside in a separate contract

jeisses committed:
Refactor network e2e test to vaccount

jeisses committed:
Move vaccount's require_auth to utility function

This function will go in a shared header later so it can be imported by other
contracts.

Some strange behaviour occurs when `require_sig` is called from within
`require_auth`. There are frequent `iterator out of bounds` errors in the smart
contract thrown by the `require_sig` and `pub2addr` functions. The cause of this
is the std::get<...> calls on the variants. In the end making `require_sig` an
inline function fixes the issue, but I honestly don't know why this is
necessary.

jeisses committed:
Add shared header for vaccount functions

jeisses committed:
Add base of Effect Force smart contract (wip)

jeisses committed:
Add reward to campaign in Force smart contract

jeisses committed:
Add readme to Force smart contract

jeisses committed:
Convert vaccount IDs to uint32

uint32 should be enough to hold all required account rows. having the ID as a
uint32 allows us to create a composite key in the campaignjoin table of Effect Force.

jeisses committed:
Add joincampaign action to Force smart contract

jeisses committed:
Update Force contract readme with joincampaign action

jeisses committed:
Add ram payer field to force contract actions

jeisses committed:
Add reserve task action

invludes merkle proof validation and checks for repeitions

jeisses committed:
Add reservetask to force smart contract readme

jeisses committed:
Add a cleanTable helper to force smart contract

jeisses committed:
Add `num_tasks` to batch in force contract

jeisses committed:
Increment batch `tasks_done` when repetition is completed

jeisses committed:
Check for campaign join before reserving a task

jeisses committed:
Add task submission and pending payment tracking

jeisses committed:
Fix pk of submission table and pub key hash transactions in force

jeisses committed:
Check that eos account exists in vaccount open

jdheeter commented:
This virtual accounts contract looks really interesting. Is there some more documentation or implementation details? I'd like to learn more about how we could contribute to this and adopt it as well.

jeisses commented:
> This virtual accounts contract looks really interesting. Is there some more documentation or implementation details? I'd like to learn more about how we could contribute to this and adopt it as well.

Hi John! Thanks for the comment.
We're working on a new documentation site that will explain the inner workings and how to use this system in other apps.

For now the best documentation can be fond in this README and some high level info in my blog post:

- https://github.com/effectai/effect-network/tree/accounts/contracts/vaccount
- https://medium.com/effect-ai/tech-blog-effect-account-system-a3bd8dfe93d4

jdheeter commented:
Ok great thanks @jeisses 

Do you have an example of the frontend code? To construct/sign the transaction. I'm not sure how the interaction with Metamask works?

jeisses commented:
We have a new frontend app that works with MetaMask. I'll open source it this week and will link you to some examples then. 

There is a bit of code needed to construct the transactions and to get a correct signature from web3. But not so hard to abstract this away in a few helper functions. 

It would be great if we can get your feedback on this system - I wouldn't mind putting in the effort to make this broadly usable in the EOSIO ecosystem.

jeisses committed:
Do not allow signatures on EOS virtual transactions

previously a signature could be added on an EOS transaction, wich was ignored by
the smart contract. this can be abused because someone could use the relayer
account as the from_account.

the relayer should only work for EOS accounts. by not allowing sig on EOS
transactions in the smart contract we will avoid mistakes by relayers.

jdheeter commented:
Hey @jeisses would be happy to collaborate with you. Maybe let's setup a call.

jeisses committed:
Upate jungle deployment script to deploy Force to jungle 3

jeisses committed:
Fix bug where global task data has to be unique

this is because the hash of the data has to be unique in the submission
table. we now prepend the batch primary key to all task data before contructing
merkle trees, this means task data only has to be unique within a batch.

jeisses committed:
Add memo field to vtransfer action

this is useful for passing data to receiver notifications. also makes it more in
line with the eosio.token transfer action.

jeisses committed:
Implement batch funding and publishing

jeisses committed:
Only allow batch to be published once

jeisses committed:
Add account_id and sig params to publishbatch action

jeisses committed:
Fix integer overflow in batch deposits

jeisses committed:
Fix publishbatch unit tests

jeisses committed:
Account for repetitions when checking the balance in publishbatch

jeisses committed:
Fix batch funding equality check in publichbatch

jabbarn committed:
Add edit campaign action to force smart contract (#66)

* create edit-campaign logic

* Fix indent

Co-authored-by: Jesse Eisses <jesse.eisses@gmail.com>

jabbarn committed:
implement delete campaign logic, change tests

jabbarn committed:
remove unused parameters, rename methods, change tests

jeisses committed:
Improve unit tests for rmcampaign

it's good to check that the campaigns are actually removed from the tables on
the blockchain. also a good test to add is to check that someone else than the
campaign owner is not able to execute the action.

jeisses committed:
Merge pull request #68 from effectai/force-erase-campaign

Delete campaign.

jabbarn committed:
implement rmbatch logic.

jabbarn committed:
implement payout logic.

jabbarn committed:
sum up amount of payments connected to account.

jabbarn committed:
remove redundant batch_id, erase entry after summing up amount

jabbarn committed:
update e2e tests.

jabbarn committed:
implement releasetask and reclaimtask logic.

jabbarn committed:
improve payout logic

jeisses committed:
Add small optimization to vaccount open action

jeisses committed:
Require auth on the force clean action for testing

jeisses committed:
Merge pull request #76 from effectai/force-erase-campaign

Force erase campaign

jeisses committed:
Improve vaccount e2e tests

jeisses committed:
Fix releasetask action in force smart contract

The problem is that it's common to have multiple `vaccount` entries for a single
`vaddress, token_contract` combo (one for each token), so find is not really
useful as you need to find the correct row.

jabbarn committed:
check if account joined campaign before reclaiming task

jeisses committed:
Improve error handling and testing for reclaimstask

jeisses committed:
Merge pull request #75 from effectai/release-reservations

implement releasetask and reclaimtask logic.

jeisses committed:
Make payout delay configurable in force contract

and add several small fixes

jeisses committed:
Simplify payout logic in force contract

the payout action contained a lot of logic on which payouts should be paid out
for an account ID. it was looping accross all payments to check which are
elegible for payout and batching them into a single transaction. this would be a
bit more effcicient transaction wise, but it's prone to bugs (e.g. it would
already fail if there are payments with different symbols or tokens).

let's move this "smart" logic from the contract to the calling code (SDK)
instead.

jeisses committed:
Merge branch 'force' into payout

jeisses committed:
Merge branch 'force' into payout

jeisses committed:
Merge pull request #70 from effectai/payout

Payout

jeisses committed:
Fix vaccount open bug with a random print statement

there is weird behaviour in the vaccount open action; on jungle3 testnet is
allows opening of multiple rows for the same account / symbol. this should not
be possible by the code. I can't reproduce it locally either.

this print statement seems to fix the contract on jungle3. of course this
doesn't make much sense, so it would be good if we can find the true cause of
the issue (could it be in the VM or compiler?).

steps to reproduce:

```
alias cleosj="cleos --url 'http://jungle3.cryptolions.io'"

cleosj push action efxaccount11 open '[["address","4a7191a8b997c6d786a21562732ca59f8c1d3c73"], ["4,EFX",efxtoken1112],efxefxefxefx]' -p efxefxefxefx@owner

cleosj push action efxaccount11 vtransfer '[65,192,["1.0000 EFX",efxtoken1112],"",null,null]' -p efxefxefxefx@active

cleosj push action efxaccount11 open
'[["address","4a7191a8b997c6d786a21562732ca59f8c1d3c72"],
["4,EFX",efxtoken1112],efxefxefxefx]' -p efxefxefxefx@owner
```

jeisses committed:
Add test for not being able to open multiple vaccounts

this replicates the bug of having multiple vaccount rows for the same symbol if
the balance is different. for some reason this is happning on jungle3 but not locally

jeisses committed:
Merge pull request #69 from effectai/force-erase-batch

implement rmbatch logic.

jabbarn committed:
Add closebatch method for pausing batches (#78)

* remove double semicolon

* create compositekey function

* pausebatch, reopenbatch + tests

* Replace reopenbatch with publishbatch in force contract

Co-authored-by: Jesse Eisses <jesse.eisses@gmail.com>

jabbarn committed:
Allow releasing reservations after delay in force (#80)

* implement release_task delay + tests

* improve release_task logic and improve tests for it

* Update code formatting

Co-authored-by: Jesse Eisses <jesse.eisses@gmail.com>

jabbarn committed:
Implement task repetitions in force (#82)

jeisses committed:
Update jungle3 deploy code

jeisses committed:
Do not create task payout rows when there is no reward

jeisses committed:
Use xfer permission for force payout transactions

jeisses committed:
Update eos-cljs

jeisses committed:
Update eosjs

newer version is needed for eos-cljs

jeisses committed:
Add mainnet deployment script

jeisses committed:
Add missing transfer permission to force Jungle3 deployment script

jeisses committed:
Fix vtransfer permission in the force mainnet deployment

jeisses committed:
Merge pull request #64 from effectai/force

Force
